### PR TITLE
Ngtsc/smarter type emitter

### PR DESCRIPTION
**Pull request by JoostK**
See individual commits

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/typecheck/test/type_parameter_emitter_spec.ts [hunk 1]
```diff
@@ -33,16 +33,21 @@ runInEachFileSystem(() => {
...
-        return '';
+        return canEmit ? '' : null;
       }
 
-      const printer = ts.createPrinter();
+      if (!canEmit) {
+        fail('canEmit must be true when emitting succeeds');
+      }
+
+      const printer = ts.createPrinter({newLine: ts.NewLineKind.LineFeed});
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/typecheck/test/type_parameter_emitter_spec.ts:50]: Is this to help when running on Windows?

**Review by petebacondarwin — COMMENTED** [review]
From first commit message:

```
The template type checker is capable of recreating generic type bounds
in a different, ...
```

It looks like this phrase was not completed?

```
when the type
parameter emitting logic was introduces
```

typo

**Review by petebacondarwin — COMMENTED** [review]
From the second commit message:

```
Type parameters with a default clause would incorrectly be emitted into
the typecheck file using the original `ts.TypeNode` for the default
clause, such that `ts.TypeReferenceNode`s within the default clause
would be invalid.
```

I didn't quite understand this sentence. Are you saying that it is not valid to have a `ts.TypeNode` inside a `ts.TypeReferenceNode`?

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_emitter.ts [hunk 2]
```diff
@@ -10,7 +10,7 @@ import {Reference} from '../../imports';
 
 /**
  * A resolved type reference can either be a `Reference`, the original `ts.TypeReferenceNode` itself
- * or null to indicate the no reference could be resolved.
+ * or null to indicate the no reference could be resolved, or that the reference can not be emitted.
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/typecheck/src/type_emitter.ts:13]:
```suggestion
 * or null to indicate that no reference could be resolved, or that the reference can not be emitted.
```

Or even better start a new sentence?

```suggestion
 * or null.
 * A value of null indicates that no reference could be resolved or that the reference can not be emitted.
```

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_emitter.ts [hunk 3]
```diff
@@ -57,10 +57,9 @@ export function canEmitType(type: ts.TypeNode, resolver: TypeReferenceResolver):
       return false;
     }
 
-    // If the type is a reference without a owning module, consider the type not to be eligible for
-    // emitting.
-    if (reference instanceof Reference && !reference.hasOwningModuleGuess) {
-      return false;
+    // If the type is a reference, consider the type not to be eligible for emitting.
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/typecheck/src/type_emitter.ts:60]:
```suggestion
    // If the type is a reference, consider the type to be eligible for emitting.
```

???

File: packages/compiler-cli/src/ngtsc/typecheck/test/type_parameter_emitter_spec.ts [hunk 4]
```diff
@@ -116,26 +116,34 @@ runInEachFileSystem(() => {
...
-          .toThrowError('A type reference to emit must be imported from an absolute module');
+      expect(() => emit(emitter)).toThrowError('Unable to emit an unresolved reference');
     });
 
     it('cannot emit references to local declarations as nested type arguments', () => {
       const emitter = createEmitter(`
           import {NgIterable} from '@angular/core';
 
-          export class Local {};
+          class Local {};
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/typecheck/test/type_parameter_emitter_spec.ts:142]: So would this test pass if `Local` was exported?
If so, should we add a test to demonstrate that?

File: packages/compiler-cli/src/ngtsc/typecheck/test/type_parameter_emitter_spec.ts [hunk 5]
```diff
@@ -170,9 +177,8 @@ runInEachFileSystem(() => {
           export class TestClass<T extends Internal> {}`,
           additionalFiles);
 
-      expect(emitter.canEmit()).toBe(false);
-      expect(() => emit(emitter))
-          .toThrowError('A type reference to emit must be imported from an absolute module');
+      expect(emitter.canEmit()).toBe(true);
+      expect(emit(emitter)).toEqual('<T extends test.Internal>');
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/typecheck/test/type_parameter_emitter_spec.ts:181]: Where does this `test` namespace come from and how would that look in practice?

**Review by petebacondarwin — APPROVED** [review]
Nice work @JoostK.
If this PR means that the fallback is no longer necessary, should it be removed too?

File: packages/compiler-cli/src/ngtsc/typecheck/test/type_parameter_emitter_spec.ts (hunk 1 above)
Comment (JoostK) [reply packages/compiler-cli/src/ngtsc/typecheck/test/type_parameter_emitter_spec.ts:50]: Indeed. I haven't seen them fail but remembered from earlier that Windows would have used `\r\n` by default, causing assertions using `\n` to fail.

File: packages/compiler-cli/src/ngtsc/typecheck/test/type_parameter_emitter_spec.ts (hunk 4 above)
Comment (JoostK) [reply packages/compiler-cli/src/ngtsc/typecheck/test/type_parameter_emitter_spec.ts:142]: If `Local` were exported, then the type emitter would indeed have succeeded.

There's a test here: https://github.com/angular/angular/pull/42492/files#diff-66249b018abab813af0a8497c3308ede071069a30ebb22dbb4c860559ef1b406R119-R127

File: packages/compiler-cli/src/ngtsc/typecheck/test/type_parameter_emitter_spec.ts (hunk 5 above)
Comment (JoostK) [reply packages/compiler-cli/src/ngtsc/typecheck/test/type_parameter_emitter_spec.ts:181]: The `test` namespace comes from the test setup that creates a fake type reference node:

https://github.com/angular/angular/blob/51610849171c8b02aa63a9cde163dd40f0585854/packages/compiler-cli/src/ngtsc/typecheck/test/type_parameter_emitter_spec.ts#L36-L39

In the actual code, the rewritten type reference is created by  ngtsc's `ReferenceEmitter` from the type-check's environment:

https://github.com/angular/angular/blob/51610849171c8b02aa63a9cde163dd40f0585854/packages/compiler-cli/src/ngtsc/typecheck/src/environment.ts#L134-L141

**Comment by JoostK** [comment]
> I didn't quite understand this sentence. Are you saying that it is not valid to have a ts.TypeNode inside a ts.TypeReferenceNode?

I added a note to clarify that an untranslated `ts.TypeReferenceNode` would likely be invalid in the type-check file, as there won't be an import statement available in the type-check file if it isn't translated.

> If this PR means that the fallback is no longer necessary, should it be removed too?

The fallback continues to be necessary, as non-exported declarations cannot be translated by the type-emitter.

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:379325560:BASE:379325585:1623697624508:8c8b64f4).

**Comment by AndrewKushnir** [comment]
@JoostK FYI presubmits went well for the changes in this PR, but it looks like it'd need an LGTM from @alxhub.

**Comment by dylhunn** [comment]
Since this is still awaiting @alxhub review, I will go ahead and remove it from the merge queue for now. Please re-add `action: merge` when ready, or let me know if you want to proceed now.

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_emitter.ts [hunk 6]
```diff
@@ -32,13 +39,15 @@ export function canEmitType(type: ts.TypeNode, resolver: TypeReferenceResolver):
...
-      visitOtherType: () => false,
-    });
+    return visitNode(type) !== INELIGIBLE;
+  }
+
+  function visitNode(node: ts.Node): typeof INELIGIBLE|undefined {
+    if (ts.isTypeReferenceNode(node) && !canEmitTypeReference(node)) {
+      return INELIGIBLE;
+    } else {
+      return ts.forEachChild(node, visitNode);
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/typecheck/src/type_emitter.ts:49]: This is getting a bit too clever... it seems we're relying on the fact that this returns `undefined` for all _other_ kinds of type nodes, correct? At the very least, we should have a comment here explaining:

1) what kinds of nodes we want to return `INELIGIBLE` for, and why.
2) the mechanism by which eligible nodes end up getting selected.

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_emitter.ts [hunk 7]
```diff
@@ -32,13 +39,15 @@ export function canEmitType(type: ts.TypeNode, resolver: TypeReferenceResolver):
...
-      visitTypeReferenceNode: type => canEmitTypeReference(type),
-      visitArrayTypeNode: type => canEmitTypeWorker(type.elementType),
-      visitKeywordType: () => true,
-      visitLiteralType: () => true,
-      visitOtherType: () => false,
-    });
+    return visitNode(type) !== INELIGIBLE;
+  }
+
+  function visitNode(node: ts.Node): typeof INELIGIBLE|undefined {
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/typecheck/src/type_emitter.ts:45]: `typeof INELIGIBLE` here is `{}`, the empty interface type. Lots of things are assignable to it, including `false`.

If we want to make a singleton value, it'd be better to use a branded type instead:

```typescript
type INELIGIBLE = {__brand: 'ineligible'};
const INELIGIBLE: INELIGIBLE = {} as INELIGIBLE;
```

This ensures we don't accidentally create other values of this type besides the constant.

**Review by alxhub — APPROVED** [review]

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_emitter.ts (hunk 7 above)
Comment (JoostK) [reply packages/compiler-cli/src/ngtsc/typecheck/src/type_emitter.ts]: Good point, updated!

**Pull request closed without merging**